paperqa/rate_limiter.py

```diff
@@ -0,0 +1,360 @@
+            )
+            matching_keys.extend(list(keys))
+
+        await client.quit()
+
+        return [self.parse_key(key.decode()) for key in matching_keys]
+
+    def get_in_memory_limit_keys(
+        self,
+    ) -> list[tuple[RateLimitItem, tuple[str, str | MATCH_ALL_INPUTS]]]:
+        """Returns a list of current RateLimitItems with tuples of namespace and primary key."""
+        return [self.parse_key(key) for key in self.storage.events]
+
+    async def get_limit_keys(
+        self,
+    ) -> list[tuple[RateLimitItem, tuple[str, str | MATCH_ALL_INPUTS]]]:
+        if os.environ.get("REDIS_URL") and not self.use_in_memory:
+            return await self.get_rate_limit_keys()
+        return self.get_in_memory_limit_keys()
+
+    async def rate_limit_status(self):
+
+        limit_status = {}
+
+        for rate_limit, (namespace, primary_key) in await self.get_limit_keys():
+            period_start, n_items_in_period = await self.storage.get_moving_window(
+                rate_limit.key_for(*(namespace, primary_key or "")),
+                rate_limit.amount,
+                rate_limit.get_expiry(),
+            )
+            limit_status[(namespace, primary_key)] = {
+                "period_start": period_start,
+                "n_items_in_period": n_items_in_period,
+                "period_seconds": rate_limit.GRANULARITY.seconds,
+                "period_name": rate_limit.GRANULARITY.name,
+                "period_cap": rate_limit.amount,
+            }
+        return limit_status
+
+    async def try_acquire(
+        self,
+        namespace_and_key: tuple[str, str | MATCH_ALL_INPUTS],
+        rate_limit: RateLimitItem | None = None,
+        machine_id: int = 0,
+        timeout: float = GLOBAL_RATE_LIMITER_TIMEOUT,  # noqa: ASYNC109
+        weight: int = 1,
+        raise_impossible_limits: bool = False,
+    ) -> None:
+        """Returns when the limit is satisfied for the namespace_and_key.
+
+        Args:
+            namespace_and_key: is composed of a tuple with namespace (e.g. "get") and
+            a primary-key (e.g. "arxiv.org"). namespaces can be nested with multiple '|',
+            primary-keys in the "get" namespace will be stripped to the domain.
+
+            rate_limit: Optional RateLimitItem to be used for the namespace and primary-key.
+            If not provided, RATE_CONFIG will be used to find the rate limit.
+
+            machine_id: will be used to modify the namespace of GET requests if
+            the primary key is not in the NO_MACHINE_ID_EXTENSIONS list.
+            In that case, the outbound IP will be used to modify the namespace.
+
+            timeout: is the maximum time to wait for the rate limit to be satisfied.
+
+            weight: is the cost of the request, default is 1. (could be tokens for example)
+
+            raise_impossible_limits: flag will raise a ValueError for weights that exceed the rate.
```

Comment:
For args, Google docstring style is:
- Don't put newlines
- Indent multilines by 4 spaces